hls4ml/model/optimizer/passes/batchnorm_opt.py

```diff
@@ -0,0 +1,269 @@
+        )
+        return is_match
+
+    def transform(self, model, node):
+        """
+        Remove the batch norm
+        """
+        const_node = node.get_input_node(node.inputs[0])
+
+        const_prec = const_node.get_output_variable().type.precision
+
+        new_val = const_node.value * node.weights['scale'].data_unquantized + node.weights['bias'].data_unquantized
+
+        const_node.set_attr('value', new_val)
+        const_node.set_attr('quantizer', node.get_attr('quantizer'))  # None if not defined
+
+        if isinstance(node.get_output_variable().type.precision, UnspecifiedPrecisionType):
+            if isinstance(const_prec, UnspecifiedPrecisionType):
+                pass  # leave it as is
+            else:
+                const_node.get_output_variable().type.precision = UnspecifiedPrecisionType()  # default
+                # propagate precision
+                scale_q = node.get_attr('scale_quantizer')
+                bias_q = node.get_attr('bias_quantizer')
+                if scale_q and bias_q:
+                    # propagate precsion
+                    scale_prec = scale_q.hls_type
+                    bias_prec = bias_q.hls_type
+                    if scale_prec not in (IntegerPrecisionType, FixedPrecisionType) or bias_prec not in (
+                        IntegerPrecisionType,
+                        FixedPrecisionType,
+                    ):
+                        print("Warning:  output type not propagated for constant merge")
+                    else:
+                        signed_prod = const_prec.signed or scale_prec.signed
+                        w_prod = const_prec.width + scale_prec.width
+                        i_prod = const_prec.integer + scale_prec.integer
+                        signed = signed_prod or bias_prec.signed
+                        i_tot = (
+                            max(
+                                i_prod + (bias_prec.signed and not signed_prod),
+                                bias_prec.ingeter + (signed_prod and not bias_prec.signed),
+                            )
+                            + 1
+                        )
+                        w_tot = i_tot + max(w_prod - i_prod, bias_prec.width - bias_prec.integer)
+                        new_prec = FixedPrecisionType(w_tot, i_tot, signed)
+                        const_node.set_attr('quantizer', QuantNodeQuantizer(new_prec))
+                        const_node.get_output_variable().type.precision = new_prec
+        else:
+            const_node.get_output_variable().type.precision = node.get_output_variable().type.precision
+
+        # remove the batch norm node
+        model.remove_node(node, rewire=True)
+
+        return True
+
+
+class FuseConsecutiveBatchNormalization(OptimizerPass):
+    """
+    OptimizerPass to merge consecutive BatchNormalization layers,
+    only if the earlier one does not have quantization specified
```

Comment:
The code below seems to match also in the case when the current node does not have quantization specified, not just the earlier one. Does this description need to be updated?

Reply:
I may have been referring to the output variable not having a special quantization required (i.e. it's `UnspecifiedPrecisionType`). I did play around with the requirements after the fact, so I should maybe update the description.

Reply:
I do wonder if the matching is too loose. It basically is saying that if at least one bias or scale is unquantized, then the resulting merged values are unquantized. This doesn't increas the number of unquantized scales of biases.

Reply:
I really have no clue what the best configuration here would be. But I would be inclined to just keep it as is for and see if it causes problems down the road.